runtime/alloc/alloc.c

```diff
@@ -0,0 +1,138 @@
+#include <string.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <stdint.h>
+#include <stdbool.h>
+#include "jemalloc/jemalloc.h"
+
+#include "runtime/alloc.h"
+
+const size_t BLOCK_SIZE = 1024 * 1024;
+
+bool true_is_fromspace = false;
+
+static char* first_block = 0;
+static char* first_tospace_block = 0;
+static char* block = 0;
+static char* block_start = 0;
+static char* block_end = 0;
+
+char *fromspace_ptr() {
+  return first_block;
+}
+
+char *alloc_ptr() {
+  return block;
+}
+
+char *arena_ptr() {
+  return block_start;
+}
+
+void koreAllocSwap() {
+  char *tmp = first_block;
+  first_block = first_tospace_block;
+  first_tospace_block = tmp;
+  block = first_block ? first_block + sizeof(memory_block_header) : 0;
+  block_start = first_block;
+  block_end = first_block ? first_block + BLOCK_SIZE : first_block;
+  true_is_fromspace = !true_is_fromspace;
+}
+
+static void* superblock_ptr = 0;
+static unsigned blocks_left = 0;
+
+static void* megabyte_malloc() {
+  if (blocks_left == 0) {
+    blocks_left = 15;
+    posix_memalign(&superblock_ptr, BLOCK_SIZE, BLOCK_SIZE * 15);
+  }
+  blocks_left--;
+  void* result = superblock_ptr;
+  superblock_ptr += BLOCK_SIZE;
+  return result;
+}
+
+static void freshBlock() {
+    char *nextBlock;
+    if (block_start == 0) {
+      nextBlock = megabyte_malloc();
+      first_block = nextBlock;
+      memory_block_header hdr;
+      hdr.next_block = 0;
+      hdr.semispace = true_is_fromspace;
+      memcpy(nextBlock, &hdr, sizeof(hdr));
+    } else {
+      nextBlock = *(char**)block_start;
+      if (block != block_end) {
+        if (block_end - block == 8) {
+          *(uint64_t *)block = 0x0000400000000000LL; // 8 bit sentinel value
+        } else {
+          *(uint64_t *)block = block_end - block - 8; // 16-bit or more sentinel value
+        }
+      }
+      if (!nextBlock) {
+        DBG("Allocating new block for the first time in semispace %d\n", true_is_fromspace);
+        nextBlock = megabyte_malloc();
+        *(char **)block_start = nextBlock;
+        memory_block_header hdr;
+        hdr.next_block = 0;
+        hdr.semispace = true_is_fromspace;
+        memcpy(nextBlock, &hdr, sizeof(hdr));
+      }
+    }
+    block = nextBlock + sizeof(memory_block_header);
+    block_start = nextBlock;
+    block_end = nextBlock + BLOCK_SIZE;
+    DBG("New block at %p (remaining %zd)\n", block, BLOCK_SIZE - sizeof(memory_block_header));
+}
+
+static void* __attribute__ ((noinline)) doAllocSlow(size_t requested) {
+  DBG("Block at %p too small, %zd remaining but %zd needed\n", block, block_end-block, requested);
+  if (requested > BLOCK_SIZE - sizeof(memory_block_header)) {
```

Comment:
Do we have any provision for collecting these objects? Are we assuming they contain no pointers?

Reply:
we're assuming that they're strings, since those are the only objects capable of being that large. So they shouldn't have pointers. Eventually we probably just want to put these directly in the old generation, but right now they're not being collected, no.

Reply:
Could there be pointers to them that we try to visit and get block metadata from during garbage collection? I think that was covered by the part where we check a header bit in migrate_once, but you should say for sure.

Reply:
this is handled by the set_len macro in header.h. When the block header of strings are initialized, we compare on the length of the string and if it's large enough that this code path would have been exercised by the allocation, we set the NOT_YOUNG_OBJECT_BIT, which prevents us from trying to migrate the object during gc.